dilated-file-transfer.md

```diff
@@ -0,0 +1,484 @@
+# Dilated File-Transfer Protocol
+
+This version of the file-transfer protocol is a complete replacement for the original (referred to as "classic") file-transfer protocol.
+
+Both sides must support and use Dilation (see `dilation-protocol.md`).
+
+Any all-caps words ("MAY", "MUST", etc) follow RFC2119 conventions.
+
+    NOTE: there are several open questions / discussion points, some with corresponding "XXX" comments inline. See also [Discussion and Open Questions](#discussion)
+
+
+## Philosophic Note
+
+The place to be opinionated is in client implementations.
+The place to be flexible is in protocol specifications.
+
+Everyone reading this likely has a particular sort of user experience (UX) in mind; the protocol should _allow_ reasonable features but shouldn't _demand_ any particular UX.
+
+The protocol absolutely MUST be extensible in the future (we can't do everything correctly immediately).
+
+
+## Overview and Features
+
+This specification is an application-level Magic Wormhole protocol defining a flexible, "session"-based approach to file transfer.
+Dilation must be supported by both clients (see :ref:`dilation-protocol.md`).
+Client implementations can allow either side to offer files/directories to send while the other side may accept or reject each offer.
+Either side MAY terminate the transfer session (by closing the wormhole).
+Either side MAY select a one-way mode, similar to the classic protocol.
+
+Files or directories are offered and sent individually, with no dependency on zip or other archive formats.
+
+Metadata is included in the offers to allow the receiver to decide if they want that file before the transfer begins.
+
+"Offers" generally correspond to what a user might select; a single-file offer is possible but so is a directory.
+In both cases, they are treated as "an offer" although a directory may consist of dozens or more individual files.
+
+
+## Version Negotiation
+
+There is an existing file-transfer protocol which does not use Dilation (called "classic" in this document).
+Clients supporting newer versions of file-transfer (i.e. the one in this document) SHOULD offer backwards compatibility.
+
+In the mailbox protocol, applications can indicate version information.
+The classic file-transfer protocol doesn't use this feature so the version information is empty.
+This new protocol will include a dict like:
+
+```json
+{
+    "transfer": {
```

Comment:
I suggest renaming this into `dilated-transfer`, and also define a field for classic transfer. Especially if clients are not required to supported the latter (as per line 41 above), this should be communicated.

Maybe move the version number field into the protocol identifier, like `dilated-transfer-v1`. This will allow us to change the field structure in future versions without getting into trouble with parsing. (This also aligns with the comment in line 61)

I think "features" is called "abilities" in other Magic Wormhole protocols, so maybe use that for consistency. Also, the "core" protocol should just be implied, so the default is an empty list.

Reply:
Well, the _lack_ of a field indicates classic transfer (although I'm definitely not opposed to re-naming this one to be more specific) so I'm not sure what utility there is in defining a second way to identify "classic" (since we _have_ to always take the absence of any information to be "classic", since that's what all current implementations do).

Other tweaks sound fine to me.

Reply:
Consider a case where we have versions 1, 2 and 3; now a conforming v3 implementation with backwards-compatibility will have to include three (probably very similar) `dicts` in the versioning information: `"transfer-v1"`, `"transfer-v2"` and `"transfer-v3"`

This is (partly) where the _list of versions_ idea comes from; an implementation can then say "I support v1 and v3" whereas with a simple `int` it cannot do that. Nevertheless, each side does need an unambiguous way to decide _which_ single version to use.

Burning the version into the key feels very similar to the "list" method, but as you say has the advantage that existing fields can have their values (or the interpretation of them) changed.

Reply:
Okay, I changed the version information to be in the key so we have a list now .. more in line with Dilation negotiation too. That is, `"transfer-v1"`, `"transver-v2"` etc.

Kept `"transfer-"` because it's shorter and we don't need to distinguish other (non-dilated) transfer.

Reply:
I'm keeping "features" for now, because I think it actually _reduces_ confusion; the thing that uses `"abilities"` is the _wormhole protocol level_ feature negotiation and since distinguishing that from the "app_versions" is already a little tricky / confusing, _not_ overlapping "abilities" here too makes it easier for readers to distinguish one from the other.